projects/app_update_out_of_date.kp/knowledge.md

```diff
@@ -0,0 +1,1067 @@
+Get the settings based on dates.
+
+
+```python
+if runManually is None:
+    todayEnvStr = environ.get("date", None)
+
+assert (todayEnvStr is not None), "The date environment parameter is missing."
+today = dt.datetime.strptime(todayEnvStr, "%Y%m%d").date()
+
+# MON = 0, SAT = 5, SUN = 6 -> SUN = 0, MON = 1, SAT = 6
+dayIndex = (today.weekday() + 1) % 7
+# Filename used to save the report's JSON
+reportFilename = (today - datetime.timedelta(dayIndex)).strftime("%Y%m%d")
+# Maximum report date which is the previous Saturday
+maxReportDate = today - datetime.timedelta(7 + dayIndex - 6)
+# Suffix of the longitudinal datasource name to use
+longitudinalSuffix = maxReportDate.strftime("%Y%m%d")
+# String used in the SQL queries to limit records to the maximum report date.
+# Since the queries use less than this is the day after the previous Saturday.
+maxReportDateSQL = (maxReportDate + dt.timedelta(days=1)).strftime("%Y-%m-%d")
+# The Sunday prior to the last Saturday
+minReportDate = maxReportDate - dt.timedelta(days=6)
+# String used in the SQL queries to limit records to the minimum report date
+# Since the queries use greater than this is six days prior to the previous Saturday.
+minReportDateSQL = minReportDate.strftime("%Y-%m-%d")
+# Date used to limit records to the number of weeks specified by
+# weeksOfSubsessionData prior to the maximum report date
+minSubsessionDate = maxReportDate - dt.timedelta(weeks=weeksOfSubsessionData)
+# Date used to compute the latest version from firefox_history_major_releases.json
+latestVerDateStr = (maxReportDate - dt.timedelta(days=7)).strftime("%Y-%m-%d")
+
+print "maxReportDate     : " + maxReportDate.strftime("%Y%m%d")
+print "maxReportDateSQL  : " + maxReportDateSQL
+print "minReportDate     : " + minReportDate.strftime("%Y%m%d")
+print "minReportDateSQL  : " + minReportDateSQL
+print "minSubsessionDate : " + minSubsessionDate.strftime("%Y%m%d")
+print "reportFilename    : " + reportFilename
+print "latestVerDateStr  : " + latestVerDateStr
+```
+Get the latest Firefox version available based on the date.
+
+
+```python
+def latestVersionOnDate(date, majorReleases):
+    latestDate = u"1900-01-01"
+    latestVer = 0
+    for version, releaseDate in majorReleases.iteritems():
+        versionInt = int(version.split(".")[0])
+        if releaseDate <= date and releaseDate >= latestDate and versionInt >= latestVer:
+            latestDate = releaseDate
+            latestVer = versionInt
+
+    return latestVer
+
+majorReleasesJSON = urllib2.urlopen("https://product-details.mozilla.org/1.0/firefox_history_major_releases.json").read()
+majorReleases = json.loads(majorReleasesJSON)
+latestVersion = latestVersionOnDate(latestVerDateStr, majorReleases)
+earliestUpToDateVersion = str(latestVersion - upToDateReleases)
+
+print "Latest Version: " + str(latestVersion)
+```
+Create a dictionary to store the general settings that will be written to a JSON file.
+
+
+```python
+reportDetailsDict = defaultdict(int)
```

Comment:
I just went with using the same type of dict as countByKey and friends returned without thinking about it. I changed it to use a standard dict.